sys/riotboot/flashwrite.c

```diff
@@ -73,22 +98,52 @@ int riotboot_flashwrite_putbytes(riotboot_flashwrite_t *state,
 
     while (len) {
         size_t flashpage_pos = state->offset % FLASHPAGE_SIZE;
-        size_t flashpage_avail = FLASHPAGE_SIZE - flashpage_pos;
+        size_t flashwrite_buffer_pos = state->offset % RIOTBOOT_FLASHPAGE_BUFFER_SIZE;
+        size_t flashpage_avail = RIOTBOOT_FLASHPAGE_BUFFER_SIZE - flashwrite_buffer_pos;
 
         size_t to_copy = min(flashpage_avail, len);
 
-        memcpy(state->flashpage_buf + flashpage_pos, bytes, to_copy);
+        if (CONFIG_RIOTBOOT_FLASHWRITE_RAW && flashpage_pos == 0) {
+            /* Erase the next page */
+            state->flashpage++;
+            flashpage_write(state->flashpage, NULL);
+        }
+        if (CONFIG_RIOTBOOT_FLASHWRITE_RAW &&
+                flashwrite_buffer_pos == 0) {
+            memset(state->flashpage_buf, 0, RIOTBOOT_FLASHPAGE_BUFFER_SIZE);
+        };
+
+        memcpy(state->flashpage_buf + flashwrite_buffer_pos, bytes, to_copy);
         flashpage_avail -= to_copy;
 
+
         state->offset += to_copy;
         bytes += to_copy;
         len -= to_copy;
         if ((!flashpage_avail) || (!more)) {
-            if (flashpage_write_and_verify(state->flashpage, state->flashpage_buf) != FLASHPAGE_OK) {
+#if CONFIG_RIOTBOOT_FLASHWRITE_RAW  /* Guards access to state::firstblock_buf */
+            void * addr = flashpage_addr(state->flashpage);
+            if (addr == riotboot_slot_get_hdr(state->target_slot) &&
+                    state->offset == RIOTBOOT_FLASHPAGE_BUFFER_SIZE) {
+                /* Skip flashing the first block, store it for later to flash it
+                 * during the flashwrite_finish function */
+                memcpy(state->firstblock_buf,
+                       state->flashpage_buf, RIOTBOOT_FLASHPAGE_BUFFER_SIZE);
+            }
+            else {
+                flashpage_write_raw((uint8_t*)addr + flashpage_pos,
+                                    state->flashpage_buf,
+                                    RIOTBOOT_FLASHPAGE_BUFFER_SIZE);
+            }
```

Comment:
How does this respect BLOCK_ALIGNMENT?  If the required alignment is 8 bytes then when the offset matches `RIOTBOOT_MAGIC` then `state->offset == RIOTBOOT_FLASHPAGE_BUFFER_SIZE` will not enter and then an unaligned write attempt will happen.

Reply:
I think we should handle that in a common place.
I added code to the sam0 flashpage driver to handle [unaligned writes](https://github.com/RIOT-OS/RIOT/blob/master/cpu/sam0_common/periph/flashpage.c#L160), this could be used by other implementations too.